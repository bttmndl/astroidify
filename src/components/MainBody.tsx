import React, {useState, useEffect} from 'react'
import "./styles.css";
import axios from 'axios';
import LoadingPage from './LoadingPage';

type AstroidData = {
    id : string,
    name : string,
    nasa_jpl_url: string,
    is_potentially_hazardous_asteroid: boolean,
};

interface Props {
  showLoader: boolean,
  setShowLoader: React.Dispatch<React.SetStateAction<boolean>>,
  inputValue : string,
}

const API_KEY = "2f5IdnsL4eoPjdcERC1vvB1rbF8VDq5Deh4cc2XQ";


const MainBody: React.FC<Props> = ({showLoader, setShowLoader, inputValue}) => {
    const [astroidData, setAstroidData] = useState<AstroidData | null>(null);
    const [error, setError] = useState<string | null>(null);
    

    //fetching the particular astroid data which user typed 
    useEffect(() => {
        async function getData() {
            try {
                const response = await axios.get(
                `https://api.nasa.gov/neo/rest/v1/neo/${inputValue}?api_key=${API_KEY}`
                );
                setAstroidData(response.data);
                setShowLoader(false);
                setError(null);
            } catch (error) {
                console.error(error);
                setError("An error occurred while fetching the asteroid data.");
                setShowLoader(false);
            }
        }

        if(inputValue)getData();

    }, [inputValue]);

    return (
        <>
            <br /><br /><br />

            {
                showLoader ? <LoadingPage/> :
                    astroidData && 
                    <div style={{width:"100%"}} className='mainBody'>
                        <div className='mainBody_inner'>
                            <h1><span className='innerMainspan'>Astroid Name</span></h1>
                            <h2 className='innerMainspanValue'>{astroidData?.name}</h2>
                        </div>
                        <div className='mainBody_inner'>
                            <h1><span className='innerMainspan'>Find More Info </span></h1>
                            <a href={astroidData?.nasa_jpl_url} target="_blank"/> <h2 className='innerMainspanValue'>{astroidData?.nasa_jpl_url}</h2>
                        </div>
                        <div className='mainBody_inner'>
                            <h1><span className='innerMainspan'>Is this Asotroid Hazardous?</span></h1>
                            <h2 className='innerMainspanValue'>{astroidData?.is_potentially_hazardous_asteroid?(<span>YES</span>):(<span>NO</span>)}</h2>
                        </div>
                    </div>
                
            }
        </>
    );
}


export default MainBody