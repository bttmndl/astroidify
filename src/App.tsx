import React, { useState } from 'react';
import './App.css';
import InputField from "./components/InputField";
import MainBody from "./components/MainBody";
import RandomButton from './components/RandomButton';

const App: React.FC =  ()=> {
  //taking the input from user input
  const [inputValue, setInputValue] = useState<string>("");
  const [flag, setFlag] = useState<boolean>(false);
  const [showLoader, setShowLoader] = useState<boolean>(false);
  
  return (
    <div className="App">
      <span className="heading">Astroidify</span>

      <InputField setInputValue={setInputValue}/>
      <RandomButton flag={flag} setFlag={setFlag} setShowLoader={setShowLoader} setInputValue={setInputValue} />
      <MainBody showLoader={showLoader} setShowLoader={setShowLoader} inputValue ={inputValue}/>
      <h1>1st</h1>
    </div>
  );
}

export default App;
