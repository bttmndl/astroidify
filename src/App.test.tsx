import {
  fireEvent,
  getByLabelText,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { rest } from 'msw';
import renderer from "react-test-renderer";
import App from "./App";
import { setupServer } from 'msw/node';
import MainBody from './components/MainBody';

describe("App component snapshot test UI", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});



const API_KEY = "2f5IdnsL4eoPjdcERC1vvB1rbF8VDq5Deh4cc2XQ";

const handlers = [
  rest.get(`https://api.nasa.gov/neo/rest/v1/neo/12345?api_key=${API_KEY}`, (req, res, ctx) => {
    return res(ctx.json({
      id: "12345",
      name: "Mock Astroid",
      nasa_jpl_url: "https://www.nasa.gov/asteroids-comets-timeline#/map",
      is_potentially_hazardous_asteroid: true
    }))
  })
];


const server = setupServer(
  rest.get('https://api.nasa.gov/neo/rest/v1/neo/12345?api_key=2f5IdnsL4eoPjdcERC1vvB1rbF8VDq5Deh4cc2XQ', (req, res, ctx) => {
    return res(ctx.json({
      id: "3542519",
      name: "(2010 PK9)",
      nasa_jpl_url: "http://ssd.jpl.nasa.gov/sbdb.cgi?sstr=3542519",
      is_potentially_hazardous_asteroid: true
    }))
  })
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('MainBody', () => {
  test('renders the asteroid data', async () => {
    render(<MainBody showLoader={false} setShowLoader={jest.fn()} inputValue="12345" />);

    const asteroidName = await screen.findByText(/Mock Astroid/i);
    const asteroidUrl = await screen.findByText(/https:\/\/www.nasa.gov\/asteroids-comets-timeline#\/map/i);
    const isHazardous = await screen.findByText(/YES/i);

    expect(asteroidName).toBeInTheDocument();
    expect(asteroidUrl).toBeInTheDocument();
    expect(isHazardous).toBeInTheDocument();
  });
});








