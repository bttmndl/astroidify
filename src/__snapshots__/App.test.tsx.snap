// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component snapshot test UI renders correctly 1`] = `
<div
  className="App"
>
  <span
    className="heading"
  >
    Astroidify
  </span>
  <form
    className="input"
  >
    <input
      className="input_box"
      onChange={[Function]}
      placeholder="Enter Astroid Id"
      type="text"
    />
    <button
      className="input_button"
      disabled={true}
      onClick={[Function]}
    >
      Find
    </button>
  </form>
  <button
    className="input_button_random"
    onClick={[Function]}
  >
    Random Astroid
  </button>
  <br />
  <br />
  <br />
</div>
`;

exports[`App component test renders correctly 1`] = `
<div
  className="App"
>
  <span
    className="heading"
  >
    Astroidify
  </span>
  <form
    className="input"
  >
    <input
      className="input_box"
      onChange={[Function]}
      placeholder="Enter Astroid Id"
      type="text"
    />
    <button
      className="input_button"
      disabled={true}
      onClick={[Function]}
    >
      Find
    </button>
  </form>
  <button
    className="input_button_random"
    onClick={[Function]}
  >
    Random Astroid
  </button>
  <br />
  <br />
  <br />
</div>
`;

exports[`Country List Data renders correctly 1`] = `
<div
  className="App"
>
  <span
    className="heading"
  >
    Astroidify
  </span>
  <form
    className="input"
  >
    <input
      className="input_box"
      onChange={[Function]}
      placeholder="Enter Astroid Id"
      type="text"
    />
    <button
      className="input_button"
      disabled={true}
      onClick={[Function]}
    >
      Find
    </button>
  </form>
  <button
    className="input_button_random"
    onClick={[Function]}
  >
    Random Astroid
  </button>
  <br />
  <br />
  <br />
</div>
`;
